### ISSUE TITLE
NTP time synchronization for parallel movements of multiple motors

### ISSUE THREAD

**Issue by Neumi** [Neumi/ethersweep#14]
Use NTP (or similar) synchronized time to run actions at before defined time points.

**Comment by Duckle29** [comment]
I just saw this pop up on youtube and it's interesting.

I think you'd want to use a setup that supports [IEEE 1588 PTP](https://en.wikipedia.org/wiki/Precision_Time_Protocol) which is a way to synchronize clocks on a network to sub-microsecond level.

It requires a phy that supports it, and I'm unsure if the wiznet phy+mac supports it. 

One approach could be a new version of this that uses an [STM32F107xx](https://www.st.com/en/microcontrollers-microprocessors/stm32f107vc.html) which has a built in Ethernet MAC with 1588 support. (However the support seems weirdly undocumented from [a quick google](https://community.st.com/t5/stm32-mcus-embedded-software/where-is-the-stm32f107xx-ieee-1588-ptpd-package/m-p/343933))

I'm unsure if the phy also needs support for this, but something like the Ti [DP83640](https://www.ti.com/lit/ds/symlink/dp83640.pdf) Does implement it. ([Ti AN-1729](https://www.ti.com/lit/an/snla099d/snla099d.pdf))

**Comment by Duckle29** [comment]
I did some more digging on the F107, and it does seem like it'd be a fair bit of work to implement, but seems fun. I found [This TCP/IP stack](https://andybrown.me.uk/2013/07/14/stm32plusnet/) that supports it

**Comment by Neumi** [comment]
This is some valuable information. I also thought a bit about using an STM with built-in ethernet, but in the end didn't find a good solution to the hardware layout. The W5500 sits on the top board and only SPI has to go through the pin headers. Sending the ethernet lines through them wouldn't work I suppose (+ EMI/EMC problems). The other option would be to have the uC on the top board as well and route everything else down. But then the space on the top board is pretty much gone. 
But would be worth a try! I always wanted to play around with the built in Ethernet and CAN stuff anyway! :D 
Replacing the W5500 would of course be no problem. I choose it because it is widely used, has good drivers, is cheap and everywhere in stock. (I also didn't want to take a deep dive into networking + physical layer stuff :D) I might make use of the hardware interrupt pin (pin 36 on W5500) to get better packet timing. This is currently just connected, but not used in software as for now. This might help with some latency issues (at least the latency that originates from the uC code doing something else ;)).

**Comment by Duckle29** [comment]
Yeah, honestly doing the digging on the F107 it seemed cool.... and a LOT of work :sweat_smile:

**Comment by Neumi** [comment]
I just checked the F107.... It would be very hard to get them on one board without double-sided assembly. And even with double-sided assembly the THT RJ45 and the F107 would not fit side by side on the board. But I would like to keep the THT RJ45 for stability reasons... The F107 with a smaller form factor like some BGA would be great :D

**Comment by PhilCowley** [comment]
To get synchronised movement, could you not try a two stage solution. First send the requested position/speed etc. to each motor, then, once all the motors are ready, send a broadcast "go" signal? This would remove any requirement for an NTP server on the network and any requirements to update RTCs in the controllers.

**Comment by Duckle29** [comment]
I haven't really looked in to this more since, but I did brush by some comparisons of hardware PtP like that (very expensive) Ti phy does, compared to software only PtP.

Might be much more interesting to see what kind of sync you can achieve with just a software solution? 
I found this: https://github.com/mpthompson/stm32_ptpd

**Comment by asteppke** [comment]
Just as a reference for the discussion: With a direct Ethernet connection without special hardware and using NTP one can expect synchronization accuracies below 1ms. For many motion control tasks this is already sufficient. Servo PID loops for example often run with 1 kHz. 

With PTP in a pure software mode one can achieve accuracies below 10us, and with dedicated hardware below 100ns. 

The effort of a software NTP client implementation is quite small, even from scratch. A dedicated hardware approach looks like overkill here. If there is an easy to port PTP software client that might be beneficial but NTP would cover a lot of potential use cases.
